File: internal/cli/digest_test.go
package cli

import (
	"bytes"
	"encoding/json"
	"strings"
	"testing"

	"github.com/srajanpathak/agentctl/internal/digest"
)

func sampleDigest() *digest.Digest {
	return &digest.Digest{
		Summary: "Refactored the parser and added tests.",
		Branch:  "feature/x",
		Turns:   12,
		Status:  "idle",
		Task:    "Refactor parser",
		Files: []digest.FileChange{
			{Path: "parse.go", Added: 40, Removed: 12, Edited: true},
			{Path: "fmt.go", Added: 2, Removed: 0, Edited: false},
		},
	}
}

func TestFormatDigestHuman(t *testing.T) {
	out := formatDigest(sampleDigest())
	for _, want := range []string{
		"Refactored the parser", "feature/x", "idle", "12",
		"parse.go", "+40", "-12", "fmt.go",
	} {
		if !strings.Contains(out, want) {
			t.Errorf("human output missing %q in:\n%s", want, out)
		}
	}
}

func TestFormatDigestNoFiles(t *testing.T) {
	out := formatDigest(&digest.Digest{Summary: "Nothing changed yet.", Status: "working"})
	if !strings.Contains(out, "no files") {
		t.Errorf("want a 'no files' line, got:\n%s", out)
	}
}

func TestDigestJSONRoundTrips(t *testing.T) {
	var buf bytes.Buffer
	if err := printJSON(&buf, sampleDigest()); err != nil {
		t.Fatal(err)
	}
	var back digest.Digest
	if err := json.Unmarshal(buf.Bytes(), &back); err != nil {
		t.Fatalf("not valid JSON: %v\n%s", err, buf.String())
	}
	if back.Branch != "feature/x" || len(back.Files) != 2 {
		t.Errorf("round-trip lost data: %+v", back)
	}
}

File: web/src/lib/digest.test.ts
import { describe, it, expect } from 'vitest';
import { fileLabel, hasFiles } from './digest';
import type { Digest } from './types';

const base: Digest = {
  summary: 's', files: [], branch: 'main', turns: 1, status: 'idle', task: 't',
};

describe('digest formatting', () => {
  it('formats a file with +/- and edit marker', () => {
    expect(fileLabel({ path: 'a.go', added: 3, removed: 1, edited: true }))
      .toBe('* a.go  +3 -1');
  });
  it('uses a space marker for non-edited (git-only) files', () => {
    expect(fileLabel({ path: 'b.go', added: 0, removed: 2, edited: false }))
      .toBe('  b.go  +0 -2');
  });
  it('hasFiles is false for null or empty', () => {
    expect(hasFiles({ ...base, files: null })).toBe(false);
    expect(hasFiles({ ...base, files: [] })).toBe(false);
    expect(hasFiles({ ...base, files: [{ path: 'x', added: 0, removed: 0, edited: true }] })).toBe(true);
  });
});
